### PR TITLE
fix(opensearch): show search results only if user has access permission to the index 

### PR DESCRIPTION
**Pull request by tsdk02**
## Type of Change

- [x] Bugfix
- [ ] New feature
- [x] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

/search and /search/{domain} are currently checking for analytics ACL before hitting opensearch and returning results.

Additionally we have to check access permissions per index:

- payment attempts/intents => PaymentRead | PaymentWrite
- refunds => RefundRead | RefundWrite
- disputes => DisputeRead | DisputeWrite

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

Checking access permissions to an index for a user is necessary to show results appropriately.

## How did you test it?

- Open the dashboard locally and hit the get_global_search_results API through the global search feature (command+k).
- Search using a query (eg: USD), to get the results corresponding to respective indexes (if user has access permission)
- Evaluate and verify the results

Screenshot 2024-06-24 at 2 09 09 PM

Screenshot 2024-06-25 at 11 26 25 PM

Screenshot 2024-06-24 at 2 09 42 PM

Screenshot 2024-06-25 at 11 26 44 PM

Alternatively, it can also be verified through Postman:
- Hit the /search and /search/{domain} API's to retrieve the data
Screenshot 2024-06-24 at 1 45 52 PM
Screenshot 2024-06-24 at 2 16 30 PM

## Checklist

- [ ] I formatted the code `cargo +nightly fmt --all`
- [ ] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/router/src/services/authentication.rs
Comment (lsampras) [inline crates/router/src/services/authentication.rs:1]: @racnan Can you verify this file changes

**Review by lsampras — APPROVED** [review]
Looks good

**Review by lsampras — APPROVED** [review]

File: crates/router/src/analytics.rs [hunk 1]
```diff
@@ -653,11 +653,42 @@ pub mod routes {
...
-            |state, auth: AuthenticationData, req, _| async move {
+            |state, auth: UserFromToken, req, _| async move {
+                let role_id = auth.role_id;
+                let role_info =
+                    RoleInfo::from_role_id(&state, &role_id, &auth.merchant_id, &auth.org_id)
+                        .await
+                        .change_context(UserErrors::InternalServerError)
+                        .change_context(OpenSearchError::UnknownError)?;
+                let permissions = role_info.get_permissions_set();
+                let accessible_indexes: Vec<_> = vec![
```
Comment (ivor-juspay) [inline crates/router/src/analytics.rs:664]: can this permission vector be kept in a common place for both global_search and index_search?

File: crates/router/src/analytics.rs [hunk 2]
```diff
@@ -653,11 +653,42 @@ pub mod routes {
...
+                .filter(|(_, perm)| perm.iter().any(|p| permissions.contains(p)))
+                .map(|i| i.0)
+                .collect();
+
                 analytics::search::msearch_results(
                     &state.opensearch_client,
                     req,
-                    &auth.merchant_account.merchant_id,
+                    &auth.merchant_id,
+                    accessible_indexes,
```
Comment (ivor-juspay) [inline crates/router/src/analytics.rs:691]: what happens when user has neither of the permissions? accessible_indexes = vec![]?, will msearch run on this?

File: crates/router/src/analytics.rs (hunk 1 above)
Comment (tsdk02) [reply crates/router/src/analytics.rs]: Yes, moved it to a consts file.

File: crates/router/src/analytics.rs (hunk 2 above)
Comment (tsdk02) [reply crates/router/src/analytics.rs:675]: Resolved this by changing the `OpenMserchOutput` struct to handle cases when the user has none of the permissions to access indexes.

**Review by lsampras — DISMISSED** [review]

File: crates/router/src/analytics.rs (hunk 2 above)
Comment (ivor-juspay) [reply crates/router/src/analytics.rs:675]: can you share the the output of this test case

Comment (tsdk02) [reply crates/router/src/analytics.rs:675]: Screenshot 2024-06-26 at 1 50 44 PM
Screenshot 2024-06-26 at 1 50 08 PM

File: crates/router/src/analytics.rs [hunk 3]
```diff
@@ -674,24 +689,33 @@ pub mod routes {
...
+                let _ = OPENSEARCH_INDEX_PERMISSIONS
+                    .iter()
+                    .filter(|(ind, _)| *ind == index)
+                    .find(|i| i.1.iter().any(|p| permissions.contains(p)))
+                    .ok_or(OpenSearchError::IndexAccessNotPermittedError(index))?;
+                analytics::search::search_results(&state.opensearch_client, req, &auth.merchant_id)
+                    .await
+                    .map(ApplicationResponse::Json)
             },
             &auth::JWTAuth(Permission::Analytics),
```
Comment (ivor-juspay) [inline crates/router/src/analytics.rs:720]: is Analytics permission also required?

Comment (tsdk02) [reply crates/router/src/analytics.rs:720]: Discussed with @lsampras and he suggested to check the Analytics Permission also as of now

**Review by ivor-juspay — APPROVED** [review]

**Review by lsampras — APPROVED** [review]

**Pull request merged**